### ISSUE TITLE
Rule: don't parse date strings using `new Date` or `Date.parse`

### ISSUE THREAD

**Issue by c089** [eslint/eslint#5809]
I would like to see a rule `no-date-parse` (better name welcome) that will prevent developers from passing strings to the Date constructor or explicitly using `Date.parse`.

The following would be considered errors:

```
const date = new Date('2012-01-01');
const date = Date.parse('2012-01-01')
```

The following would not be considered errors:

```
const date = new Date(2012, 0, 1);
```

This prevents errors resulting from different implementatins across browsers. Using this to parse date strings is "strongly discouraged" as per the [MDN documentation on `Date`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Date):

> Note: parsing of date strings with the Date constructor (and Date.parse, they are equivalent) is strongly discouraged due to browser differences and inconsistencies.

**Comment by ilyavolodin** [comment]
I'm not sure this is a good fit for the core rule. But just to clarify, should a code below be flagged by this rule?
```js
const date = new Date(someDateString);
```

**Comment by c089** [comment]
If we can detect the type of that variable, yes.

**Comment by ilyavolodin** [comment]
No we can't detect the type of variable passed in. Static analysis is limited in that regard. So in that case, the only thing that would be flagged by this rule is Date constructor or `Date.parse` with a string parameter?

**Comment by c089** [comment]
All of Date.parse, but only string literals for the ctor. So bad would be:

```
new Date('anystring');
Date.parse('anystring');
Date.parse(anyValue);
```

Okay would be:
```
new Date(1); // ...
new Date(1,2) // ...and so on
new Date(anyValue);
```

**Comment by alberto** [comment]
Why would you allow
```
new Date(anyvalue)
```

but discourage
```
Date.parse(anyvalue)
```
?

**Comment by c089** [comment]
Yes, e.g. current stable version of Chrome parses `2012-01-01 00:00:00` while Safari returns NaN.

**Comment by nzakas** [comment]
I'm sorry, this doesn't fit the [criteria](http://eslint.org/docs/developer-guide/contributing/new-rules) for new core rules. If you still want the rule, you can always create a [custom rule](http://eslint.org/docs/developer-guide/working-with-rules) instead.
